File: bitcoin-stockmem-ts/src/llm/response-parser.ts
/**
 * Zod schemas for validating LLM JSON responses.
 */

import { z } from "zod";

// Step 1: Event extraction
export const ExtractedEventSchema = z.object({
  event_group: z.string(),
  event_type: z.string(),
  time: z.string().default("unknown"),
  location: z.string().default("global"),
  entities: z.array(z.string()).default([]),
  industries: z.array(z.string()).default([]),
  description: z.string(),
  extended_attrs: z.record(z.unknown()).optional().default({}),
});
export type ExtractedEvent = z.infer<typeof ExtractedEventSchema>;

// Step 2: Event merging
export const MergedEventSchema = z.object({
  event_group: z.string(),
  event_type: z.string(),
  time: z.string().default("unknown"),
  location: z.string().default("global"),
  entities: z.array(z.string()).default([]),
  industries: z.array(z.string()).default([]),
  description: z.string(),
  source_event_ids: z.array(z.number()).default([]),
});
export type MergedEvent = z.infer<typeof MergedEventSchema>;

// Step 3: Event tracking
export const TrackResultSchema = z.object({
  has_predecessor: z.boolean(),
  predecessor_id: z.number().nullable().default(null),
  delta_info: z.string().nullable().default(null),
});
export type TrackResult = z.infer<typeof TrackResultSchema>;

// Step 4: Reflection
export const ReasonResultSchema = z.object({
  "Reason for price movement": z.string(),
  "Events causing the impact": z.string(),
});
export type ReasonResult = z.infer<typeof ReasonResultSchema>;

// Step 5: Retrieval filter
export const RetrieveResultSchema = z.object({
  selected_indices: z.array(z.number()),
});
export type RetrieveResult = z.infer<typeof RetrieveResultSchema>;

// Step 6: Prediction
export const PredictResultSchema = z.object({
  "Reason for price movement": z.string(),
  "Price movement": z.string(),
});
export type PredictResult = z.infer<typeof PredictResultSchema>;

// ---------------------------------------------------------------------------
// Safe parsers (return null on failure instead of throwing)
// ---------------------------------------------------------------------------

export function parseExtractedEvents(data: unknown[]): ExtractedEvent[] {
  return data
    .map((item) => {
      try { return ExtractedEventSchema.parse(item); } catch { return null; }
    })
    .filter((x): x is ExtractedEvent => x !== null);
}

export function parseMergedEvents(data: unknown[]): MergedEvent[] {
  return data
    .map((item) => {
      try { return MergedEventSchema.parse(item); } catch { return null; }
    })
    .filter((x): x is MergedEvent => x !== null);
}

export function parseTrackResult(data: unknown): TrackResult {
  return TrackResultSchema.parse(data);
}

export function parseReasonResult(data: unknown): ReasonResult {
  return ReasonResultSchema.parse(data);
}

export function parseRetrieveResult(data: unknown): RetrieveResult {
  return RetrieveResultSchema.parse(data);
}

export function parsePredictResult(data: unknown): PredictResult {
  return PredictResultSchema.parse(data);
}

File: bitcoin_stockmem/llm/response_parser.py
"""Pydantic models for validating LLM JSON responses."""

from __future__ import annotations
from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Step 1: Event extraction
# ---------------------------------------------------------------------------
class ExtractedEvent(BaseModel):
    event_group: str
    event_type: str
    time: str = "unknown"
    location: str = "global"
    entities: list[str] = Field(default_factory=list)
    industries: list[str] = Field(default_factory=list)
    description: str
    extended_attrs: dict = Field(default_factory=dict)


# ---------------------------------------------------------------------------
# Step 2: Event merging
# ---------------------------------------------------------------------------
class MergedEvent(BaseModel):
    event_group: str
    event_type: str
    time: str = "unknown"
    location: str = "global"
    entities: list[str] = Field(default_factory=list)
    industries: list[str] = Field(default_factory=list)
    description: str
    source_event_ids: list[int] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Step 3: Event tracking
# ---------------------------------------------------------------------------
class TrackResult(BaseModel):
    has_predecessor: bool
    predecessor_id: int | None = None
    delta_info: str | None = None


# ---------------------------------------------------------------------------
# Step 4: Reflection / Reason
# ---------------------------------------------------------------------------
class ReasonResult(BaseModel):
    reason: str = Field(alias="Reason for price movement")
    key_events: str = Field(alias="Events causing the impact")


# ---------------------------------------------------------------------------
# Step 5: Retrieval filter
# ---------------------------------------------------------------------------
class RetrieveResult(BaseModel):
    selected_indices: list[int]


# ---------------------------------------------------------------------------
# Step 6: Prediction
# ---------------------------------------------------------------------------
class PredictResult(BaseModel):
    reason: str = Field(alias="Reason for price movement")
    direction: str = Field(alias="Price movement")


# ---------------------------------------------------------------------------
# Parsing helpers
# ---------------------------------------------------------------------------
def parse_extracted_events(data: list[dict]) -> list[ExtractedEvent]:
    """Parse a list of dicts into ExtractedEvent models, skipping invalid."""
    results = []
    for item in data:
        try:
            results.append(ExtractedEvent(**item))
        except Exception:
            continue
    return results


def parse_merged_events(data: list[dict]) -> list[MergedEvent]:
    results = []
    for item in data:
        try:
            results.append(MergedEvent(**item))
        except Exception:
            continue
    return results


def parse_track_result(data: dict) -> TrackResult:
    return TrackResult(**data)


def parse_reason_result(data: dict) -> ReasonResult:
    return ReasonResult(**data)


def parse_retrieve_result(data: dict) -> RetrieveResult:
    return RetrieveResult(**data)


def parse_predict_result(data: dict) -> PredictResult:
    return PredictResult(**data)
